smarts/core/lanepoints.py

```diff
@@ -0,0 +1,564 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+from __future__ import (  # to allow for typing to refer to class being defined (LinkedLanePoint)
+    annotations,
+)
+
+import math
+import queue
+from collections import defaultdict
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import List, NamedTuple
+
+import numpy as np
+
+import warnings
+
+with warnings.catch_warnings():
+    from smarts.core.utils.file import suppress_pkg_resources
+
+    warnings.filterwarnings("ignore", "numpy.ufunc size changed")
+    with suppress_pkg_resources():
+        # We force sklearn to use a different importer since sklearn's use of pkg_resources is
+        # aggressive
+        from sklearn.neighbors import KDTree
+
+from smarts.core.coordinates import Heading, Pose
+from smarts.core.utils.math import (
+    lerp,
+    radians_to_vec,
+    signed_dist_to_line,
+    squared_dist,
+    vec_2d,
+    vec_to_radians,
+)
+
+
+@dataclass(frozen=True)
+class LanePoint:
+    """a.k.a. MapPoint:  static, based only on a map, always in a Lane"""
+
+    pos: np.ndarray  # Point positioned on center of lane
+    heading: Heading  # Heading angle of lane at this point (radians)
+    lane_width: float  # Width of lane at this point (meters)
+    speed_limit: float  # Lane speed in m/s
+    lane_id: str  # ID of lane under lanepoint
+    lane_index: int  # Index of the lane this lanepoint is over. 0 is the outer(right) most lane.
+
+    def dist_to(self, p) -> float:
+        """Calculates straight line distance to the given 2D point"""
+        return np.linalg.norm(self.pos - vec_2d(p))
+
+    def relative_heading(self, h: Heading) -> Heading:
+        """Computes relative heading between the given angle and the lanepoint heading
+
+        Returns:
+            relative_heading: [-pi..pi]
+        """
+        assert isinstance(
+            h, Heading
+        ), "Heading h ({}) must be an instance of smarts.core.coordinates.Heading".format(
+            type(h)
+        )
+
+        return self.heading.relative_to(h)
+
+    def signed_lateral_error(self, p) -> float:
+        """Returns the signed lateral distance from the given point to the
+        line formed by the lanepoint position and the lanepoints heading.
+
+        Negative signals right of line and Positive left of line
+        """
+        return signed_dist_to_line(p, self.pos, self.heading.direction_vector())
+
+    @property
+    def pose(self) -> Pose:
+        return Pose.from_center(tuple(self.pos), self.heading)
+
+    def __hash__(self):
+        return hash(
+            (
+                *self.pos,
+                self.heading,
+                self.lane_width,
+                self.speed_limit,
+                self.lane_id,
+                self.lane_index,
+            )
+        )
+
+    def __eq__(self, other):
```

Comment:
Type of `other` ?

Reply:
I don't think it's appropriate to put type hints on the "other" parameter of a python `__eq__` function.  Unless we want to put `Any` (which I'm not really a fan of doing)?